packages/core/src/external/mpi/match-patients.ts

```diff
@@ -0,0 +1,147 @@
+import { intersectionWith, isEqual } from "lodash";
+import { PatientDataMPI } from "./patient-incoming-schema";
+import { normalizePatientDataMPI } from "./normalize-patient";
+import jaroWinkler from "jaro-winkler";
+
+// Define a type for the similarity function
+type SimilarityFunction = (
+  patient1: PatientDataMPI,
+  patient2: PatientDataMPI,
+  threshold: number
+) => boolean;
+
+type MatchingRule = (patient1: PatientDataMPI, patient2: PatientDataMPI) => boolean;
+
+/**
+ * `matchPatients` filters patients based on a similarity function, rules, and a threshold.
+ * @param similarityFunction - Determines if a patient is a match.
+ * @param matchingRules - Rules that determines if a patient is a match.
+ * @param patients - Array of patients.
+ * @param demo - Patient data to find matches for.
+ * @param threshold - Minimum similarity score for a match.
+ * @param greedy - If true, returns the first match. If false, returns all matches.
+ * @returns matched patients.
+ */
+export const matchPatients = (
+  similarityFunction: SimilarityFunction,
+  matchingRules: MatchingRule[],
+  patients: PatientDataMPI[],
+  demo: PatientDataMPI,
+  threshold: number,
+  greedy = true
+): PatientDataMPI[] => {
+  const matchFunction = (patient: PatientDataMPI) => {
+    const PatientDataMPI = normalizePatientDataMPI(patient);
+    if (!PatientDataMPI) {
+      return false;
+    }
+    for (const rule of matchingRules) {
+      if (rule(demo, PatientDataMPI)) {
+        return true;
+      }
+    }
+    return similarityFunction(PatientDataMPI, demo, threshold);
+  };
```

Comment:
Proabbly best to create this function outside of the function and just pass it the similarity function

Reply:
a little confused by what you mean here

Reply:
What i mean is to set the `matchFunction` outside of `matchPatients` it makes it easier to read

Reply:
I feel like that is a little overkill, since this matchFunction is already receiving two parameters, the similarityFunctions and the list of matchingRules which are themselves functions. This doesn't seem to me to be too hard to read